### PR TITLE
ADR for auditing and logging

### PR DESCRIPTION
**Pull request by Floppy**
Spelling out our proposed approach to audit logging in an ADR.

### CODE REVIEW

**Comment by Floppy** [comment]
Chris asks:

> will “* Incoming HTTP requests (though some information will only be available to Cloud Platform)” include the name of the user who made the request by default, do you know?

**Comment by benilovj** [comment]
@Floppy should we be sending PII into Kibana and AppInsights or should that be free of PII and some kind of user identifier should be used instead? If the latter, which user idenitifers should we be using so the data can be joined up at a later stage?

**Comment by Floppy** [comment]
> @Floppy should we be sending PII into Kibana and AppInsights or should that be free of PII and some kind of user identifier should be used instead? If the latter, which user identifiers should we be using so the data can be joined up at a later stage?

It should be some sort of ID that can be linked to a user, not direct PII; but I agree there's a fine line there. We should probably identify in here exactly which identifier we will use. HMPPS Auth should give us something in the OpenID Connect standard claims, but we should check.

**Comment by Floppy** [comment]
https://github.com/ministryofjustice/hmpps-auth/blob/9296135ad842e6ec01945d679666ffd46c98654a/src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/model/UserDetail.kt mentions that HMPPS Auth users each have a UUID, so that will be what we'd want to log, I think. @Mjwillis does that seem reasonable?

**Review by benilovj — APPROVED** [review]

**Comment by electrified** [comment]
> https://github.com/ministryofjustice/hmpps-auth/blob/9296135ad842e6ec01945d679666ffd46c98654a/src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/model/UserDetail.kt mentions that HMPPS Auth users each have a UUID, so that will be what we'd want to log, I think. @Mjwillis does that seem reasonable?

The UUIDs aren't exposed in the issued tokens. What we get is the Delius username and user ID, for users authenticated with NDElius.

See here for a token body:

```
{
  "sub": "EDBRINDLEY",
  "user_id": "2500084042",
  "user_name": "EDBRINDLEY",
  "scope": [
    "read"
  ],
  "auth_source": "delius",
  "name": "Ed Brindley",
  "iss": "https://sign-in-dev.hmpps.service.justice.gov.uk/auth/issuer",
  "exp": 1628710888,
  "authorities": [
    "ROLE_PREPARE_A_CASE",
    "ROLE_PROBATION"
  ],
  "jti": "u7ooUjwxp5xyEJIBzh1dAyH95TI",
  "client_id": "hmpps-manage-supervisions-dev"
}
```

**Comment by electrified** [comment]
> https://github.com/ministryofjustice/hmpps-auth/blob/9296135ad842e6ec01945d679666ffd46c98654a/src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/model/UserDetail.kt mentions that HMPPS Auth users each have a UUID, so that will be what we'd want to log, I think. @Mjwillis does that seem reasonable?

UUIDs are available in the /api/user/me response

> {
>     "username": "EDBRINDLEY",
>     "active": true,
>     "name": "Ed Brindley",
>     "authSource": "delius",
>     "userId": "2500084042",
>     "uuid": "2b261c25-d168-4330-88c1-a8c58abf6539"
> }

Just working out if you can look up a users detail based on UUID...

**Comment by Floppy** [comment]
In your example there @electrified, is the `userId` the Delius ID?

**Comment by electrified** [comment]
> In your example there @electrified, is the `userId` the Delius ID?

It is yes. As stored in the NDelius LDAP directory.

**Comment by Floppy** [comment]
OK, cool, then we want to use that UUID instead. Good to know that it's in *an* API, I'll find out if we can include it in the oauth claim so we don't have to fetch it separately.

**Comment by electrified** [comment]
> OK, cool, then we want to use that UUID instead. Good to know that it's in _an_ API, I'll find out if we can include it in the oauth claim so we don't have to fetch it separately.

It's not a UUID just an integer. 

Re the HMPPS Auth UUIDs, apparently there is an API to look up users via them, although I haven't located it myself yet.

**Comment by electrified** [comment]
> Sorry, my writing was confusing. We're right to use the `uuid` value in logs. As we said in standup, mapping that back to real people is left as an exercise for the auditor :)

Yes, although I can envision us needing to do this for incident investigation too, so we do need a mechanism.

**Comment by Floppy** [comment]
> Yes, although I can envision us needing to do this for incident investigation too, so we do need a mechanism.

Fair point. We can ask for it on their backlog.

**Comment by Floppy** [comment]
We can get user details by UUID using https://sign-in-dev.hmpps.service.justice.gov.uk/auth/swagger-ui.html#//api/authuser/getUserDetails, so that's not a problem.

File: doc/adr/0012-provide-an-audit-trail-via-logging.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+* Incoming HTTP requests (though some information will only be available to Cloud Platform)
+* API calls we make to other services such as Community API (perhaps using [Axios request interceptors](https://itnext.io/advanced-nestjs-techniques-part-2-logging-outgoing-http-requests-3c75d47c5768))
+* Any unexpected exceptions
+
+We will include the [UUID](https://github.com/ministryofjustice/hmpps-auth/blob/9296135ad842e6ec01945d679666ffd46c98654a/src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/model/UserDetail.kt) of the current authenticated user in all log messages.
+
+These log messages will be written to application `stdout` and `stderr`, which are [automatically aggregated by Cloud Platform into Azure Application Insights](https://user-guide.cloud-platform.service.justice.gov.uk/documentation/logging-an-app/log-collection-and-storage.html#application-log-collection-and-storage), and which are retained there for 13 months. Correlation IDs in AppInsights allow reconstruction of cross-service data for comprehensive audit purposes; we should test that this is working before going live.
```
Comment (Mjwillis) [inline doc/adr/0012-provide-an-audit-trail-via-logging.md:58]: We are changing App Insight to log 13 months worth of data @mattops is done this today

**Review by Mjwillis — APPROVED** [review]
LGTM

File: doc/adr/0012-provide-an-audit-trail-via-logging.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+We will include the [UUID](https://github.com/ministryofjustice/hmpps-auth/blob/9296135ad842e6ec01945d679666ffd46c98654a/src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/model/UserDetail.kt) of the current authenticated user in all log messages.
+
+In particular, we need to make sure we log all read operations done through our UI, because they will not be recorded by any other system.
+
+These log messages will be written to application `stdout` and `stderr`, which are [automatically aggregated by Cloud Platform into Azure Application Insights](https://user-guide.cloud-platform.service.justice.gov.uk/documentation/logging-an-app/log-collection-and-storage.html#application-log-collection-and-storage), and which are retained there for 13 months. Correlation IDs in AppInsights allow reconstruction of cross-service data for comprehensive audit purposes; we should test that this is working before going live.
+
+## Consequences
+
+The upcoming MLAP system requires unmodified log files to ingest; they cannot rely on AppInsights exports cannot be relied upon. When
```
Comment (andrewrlee) [inline doc/adr/0012-provide-an-audit-trail-via-logging.md:64]: ```The upcoming MLAP system requires unmodified log files to ingest; they cannot rely on AppInsights exports cannot be relied upon.```

Is a bit garbled

Comment (Floppy) [reply doc/adr/0012-provide-an-audit-trail-via-logging.md:64]: Oh yeah, thanks!

**Review by benilovj — APPROVED** [review]

File: doc/adr/0012-provide-an-audit-trail-via-logging.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+* API calls we make to other services such as Community API (perhaps using [Axios request interceptors](https://itnext.io/advanced-nestjs-techniques-part-2-logging-outgoing-http-requests-3c75d47c5768))
+* Any unexpected exceptions
+
+We will include the [UUID](https://github.com/ministryofjustice/hmpps-auth/blob/9296135ad842e6ec01945d679666ffd46c98654a/src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/model/UserDetail.kt) of the current authenticated user in all log messages.
+
+In particular, we need to make sure we log all read operations done through our UI, because they will not be recorded by any other system.
+
+These log messages will be written to application `stdout` and `stderr`, which are [automatically aggregated by Cloud Platform into Azure Application Insights](https://user-guide.cloud-platform.service.justice.gov.uk/documentation/logging-an-app/log-collection-and-storage.html#application-log-collection-and-storage), and which are retained there for 13 months. Correlation IDs in AppInsights allow reconstruction of cross-service data for comprehensive audit purposes; we should test that this is working before going live.
```
Comment (sldblog) [inline doc/adr/0012-provide-an-audit-trail-via-logging.md:60]: > automatically aggregated by Cloud Platform into Azure Application Insights

🙋‍♂️ As far as I know, shipping the log data to Application Insights has to be configured by the service itself, as part of the AppInsights client configuration.

Cloud Platform ships the logs into its own ElasticSearch cluster, where it retains it only for 30 days; I don't think they integrate with AppInsights in any way.

**Pull request merged**